File: data-structures/queues/PHP/Queue.php
<?php

include '../../../helpers/PHP/helpers.php';

class QueueNode
{
    public $data;
    public $next = null;

    public function __construct($data)
    {
        $this->data = $data;
    }
}

class Queue
{
    public $head;
    public $tail;

    public function enqueue($data)
    {
        $node = new QueueNode($data);

        if ($this->tail !== null) {
            $this->tail->next = $node;
        }

        $this->tail = $node;

        if ($this->head === null) {
            $this->head = $node;
        }
    }

    public function dequeue()
    {
        if ($this->head === null) {
            return null;
        }

        $head = $this->head->data;
        $this->head = $this->head->next;
        return $head;
    }

    public function top()
    {
        if ($this->head === null) {
            return null;
        }

        return $this->head->data;
    }

    public function isEmpty()
    {
        return $this->head === null;
    }

    public function display()
    {
        $current = $this->head;
        $count = 0;
        while ($current !== null) {
            println('- Node data: ' . $current->data);
            $current = $current->next;
            $count++;
        }
        println('-- Node count: ' . $count);
        println('');
    }
}

// Create queue
$queue = new Queue();
$queue->display();
queueAssertEqual('', $queue);

// Enqueue
println("---------- Enqueue ----------");
$queue->enqueue(42);
$queue->display();
queueAssertEqual('42', $queue);

$queue->enqueue(51);
$queue->display();
queueAssertEqual('42,51', $queue);

$queue->enqueue(11);
$queue->enqueue(455);
$queue->display();
queueAssertEqual('42,51,11,455', $queue);

// Dequeue
println("---------- Dequeue ----------");
$queue = new Queue();
$dequeue = $queue->dequeue();
queueAssertEqual('', $queue);
assertEqual(null, $dequeue);

$queue->enqueue(42);
$queue->enqueue(51);
$queue->enqueue(11);
$queue->enqueue(455);

$dequeue = $queue->dequeue();
queueAssertEqual('51,11,455', $queue);
assertEqual(42, $dequeue);

$dequeue = $queue->dequeue();
queueAssertEqual('11,455', $queue);
assertEqual(51, $dequeue);

$dequeue = $queue->dequeue();
assertEqual(11, $dequeue);

$dequeue = $queue->dequeue();
assertEqual(455, $dequeue);

$dequeue = $queue->dequeue();
assertEqual(null, $dequeue);

// Top
println("---------- Top ----------");
$queue = new Queue();
$queue->enqueue(42);
$queue->enqueue(51);

$top = $queue->top();
queueAssertEqual('42,51', $queue);
assertEqual(42, $top);

// Is empty
println("---------- Is empty ----------");
$queue = new Queue();
assertEqual(true, $queue->isEmpty());

$queue->enqueue(42);
assertEqual(false, $queue->isEmpty());

File: data-structures/queues/Java/Queue.java
public class Queue {
    private class QueueNode {
        private int data;
        private QueueNode next;

        private QueueNode(int data) {
            this.data = data;
        }
    }

    private QueueNode head;
    private QueueNode tail;

    public void enqueue(int data) {
        QueueNode node = new QueueNode(data);
        
        if (tail != null) {
            tail.next = node;
        }
        
        tail = node;

        if (head == null) {
            head = node;
        }
    }

    public int dequeue() {
        if (head == null) {
            return -1;
        }

        int data = head.data;
        head = head.next;
        if (head == null) {
            tail = null;
        }
        return data;
    }

    public int top() {
        if (head == null) {
            return -1;
        }

        return head.data;
    }

    public boolean isEmpty() {
        return head == null;
    }

    public void display() {
        if (head == null) {
            System.out.println("[]");
            return;
        }

        QueueNode current = head;
        System.out.println("[");
        while (current != null) {
            System.out.println("\t" + current.data);
            current = current.next;
        }
        System.out.println("]");
    } 

    public static void main(String[] args) {
        // Create queue
        Queue queue = new Queue();
        queue.display();

        // Enqueue
        System.out.println("---------- Enqueue ----------");
        queue.enqueue(42);
        queue.display();

        queue.enqueue(51);
        queue.display();

        queue.enqueue(11);
        queue.enqueue(455);
        queue.display();

        // Dequeue
        System.out.println("---------- Dequeue ----------");
        queue = new Queue();
        int dequeue = queue.dequeue();
        System.out.println("dequeue: " + dequeue);

        queue.enqueue(42);
        queue.enqueue(51);
        queue.enqueue(11);
        queue.enqueue(455);

        dequeue = queue.dequeue();
        System.out.println("dequeue: " + dequeue);

        dequeue = queue.dequeue();
        System.out.println("dequeue: " + dequeue);

        dequeue = queue.dequeue();
        System.out.println("dequeue: " + dequeue);

        dequeue = queue.dequeue();
        System.out.println("dequeue: " + dequeue);

        dequeue = queue.dequeue();
        System.out.println("dequeue: " + dequeue);

        // Top
        System.out.println("---------- Top ----------");
        queue = new Queue();
        int top = queue.top();
        System.out.println("top: " + top);

        queue.enqueue(42);
        queue.enqueue(51);

        top = queue.top();
        System.out.println("top: " + top);

        // Is empty
        System.out.println("---------- Is empty ----------");
        queue = new Queue();
        System.out.println(queue.isEmpty());

        queue.enqueue(42);
        System.out.println(queue.isEmpty());
    }
}